### PR TITLE
[hls] Improve DRM checks

### LINKED ISSUE #1658: ParamountPlus: SAMPLE-AES encryption is not supported yet
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2021.11.10.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

United States

### Description

Shows seem fine, however, movies can't be extracted anymore.

### Verbose log

```shell
Too long to fit here, https://paste.ee/r/RQgf1/0
```

### PR DESCRIPTION
**Pull request by nyuszika7h**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This pull request does two things:

* Report Adobe Flash Access as DRM instead of "No video formats found"
* Detect FairPlay/FlashAccess DRM when only specified HLS variant playlists rather than the master playlist (closes #1658)

### CODE REVIEW

**Comment by nyuszika7h** [comment]
Don't merge this yet, there may be an issue with the regex.

File: yt_dlp/extractor/common.py [hunk 1]
```diff
@@ -2035,10 +2035,11 @@ def _parse_m3u8_formats_and_subtitles(
...
 
-        if '#EXT-X-FAXS-CM:' in m3u8_doc:  # Adobe Flash Access
-            return formats, subtitles
-
-        has_drm = re.search(r'#EXT-X-(?:SESSION-)?KEY:.*?URI="skd://', m3u8_doc)
+        # NOTE: '#' needs to be escaped since it's a comment character in multiline regex ('\#' doesn't work)
+        has_drm = re.search(r'''(?x)(?:
+                            [#]EXT-X-FAXS-CM:|                        # Adobe Flash Access
+                            [#]EXT-X-(?:SESSION-)?KEY:.*?URI="skd://  # Apple FairPlay
+                            )''', m3u8_doc)
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:2042]:
```suggestion
        has_drm = re.search('|'.join([
            r'#EXT-X-FAXS-CM:',  # Adobe Flash Access
            r'#EXT-X-(?:SESSION-)?KEY:.*?URI="skd://',  # Apple FairPlay
        ], m3u8_doc)
```
is far more readable imo. What do u think?

File: yt_dlp/downloader/hls.py [hunk 2]
```diff
@@ -77,10 +77,19 @@ def real_download(self, filename, info_dict):
...
+                                [#]EXT-X-(?:SESSION-)?KEY:.*?URI="skd://  # Apple FairPlay
+                                )''', s)
+            if has_drm and not self.params.get('allow_unplayable_formats'):
+                self.report_error('This video is DRM protected')
+                return False
+            else:
+                message = message or 'Unsupported features have been detected'
+                fd = FFmpegFD(self.ydl, self.params)
+                self.report_warning(f'{message}; extraction will be delegated to {fd.get_basename()}')
+                return fd.real_download(filename, info_dict)
```
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:92]:
```suggestion
            message = message or 'Unsupported features have been detected'
            fd = FFmpegFD(self.ydl, self.params)
            self.report_warning(f'{message}; extraction will be delegated to {fd.get_basename()}')
            return fd.real_download(filename, info_dict)
```

File: yt_dlp/downloader/hls.py [hunk 3]
```diff
@@ -77,10 +77,19 @@ def real_download(self, filename, info_dict):
...
-            fd = FFmpegFD(self.ydl, self.params)
-            self.report_warning(f'{message}; extraction will be delegated to {fd.get_basename()}')
-            return fd.real_download(filename, info_dict)
+            # NOTE: '#' needs to be escaped since it's a comment character in multiline regex ('\#' doesn't work)
+            has_drm = re.search(r'''(?x)(?:
+                                [#]EXT-X-FAXS-CM:|                        # Adobe Flash Access
+                                [#]EXT-X-(?:SESSION-)?KEY:.*?URI="skd://  # Apple FairPlay
+                                )''', s)
+            if has_drm and not self.params.get('allow_unplayable_formats'):
+                self.report_error('This video is DRM protected')
```
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:86]:
```suggestion
                self.report_error(
                    'This video is DRM protected; Try selecting another format with --format or '
                    'add --check-formats to automatically fallback to the next best format')
```

File: yt_dlp/downloader/hls.py [hunk 4]
```diff
@@ -77,6 +77,16 @@ def real_download(self, filename, info_dict):
                 message = ('The stream has AES-128 encryption and neither ffmpeg nor pycryptodomex are available; '
                            'Decryption will be performed natively, but will be extremely slow')
         if not can_download:
+            # NOTE: '#' needs to be escaped since it's a comment character in multiline regex ('\#' doesn't work)
+            has_drm = re.search(r'''(?x)(?:
+                                [#]EXT-X-FAXS-CM:|                        # Adobe Flash Access
+                                [#]EXT-X-(?:SESSION-)?KEY:.*?URI="skd://  # Apple FairPlay
+                                )''', s)
```
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:84]:
```suggestion
            has_drm = re.search('|'.join([
                r'#EXT-X-FAXS-CM:',  # Adobe Flash Access
                r'#EXT-X-(?:SESSION-)?KEY:.*?URI="skd://',  # Apple FairPlay
            ], s)
```

File: yt_dlp/downloader/hls.py [hunk 5]
```diff
@@ -77,6 +77,15 @@ def real_download(self, filename, info_dict):
                 message = ('The stream has AES-128 encryption and neither ffmpeg nor pycryptodomex are available; '
                            'Decryption will be performed natively, but will be extremely slow')
         if not can_download:
+            has_drm = re.search('|'.join([
+                r'#EXT-X-FAXS-CM:',  # Adobe Flash Access
+                r'#EXT-X-(?:SESSION-)?KEY:.*?URI="skd://',  # Apple FairPlay
+            ], s)
```
Comment (pukkandan) [inline yt_dlp/downloader/hls.py:83]:
```suggestion
            ]), s)
```

File: yt_dlp/extractor/common.py [hunk 6]
```diff
@@ -2035,10 +2035,10 @@ def _parse_m3u8_formats_and_subtitles(
...
         formats, subtitles = [], {}
 
-        if '#EXT-X-FAXS-CM:' in m3u8_doc:  # Adobe Flash Access
-            return formats, subtitles
-
-        has_drm = re.search(r'#EXT-X-(?:SESSION-)?KEY:.*?URI="skd://', m3u8_doc)
+        has_drm = re.search('|'.join([
+            r'#EXT-X-FAXS-CM:',  # Adobe Flash Access
+            r'#EXT-X-(?:SESSION-)?KEY:.*?URI="skd://',  # Apple FairPlay
+        ], m3u8_doc)
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:2041]:
```suggestion
        ]), m3u8_doc)
```

**Review by pukkandan — APPROVED** [review]

**Pull request merged**